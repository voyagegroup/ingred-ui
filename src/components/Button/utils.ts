import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
