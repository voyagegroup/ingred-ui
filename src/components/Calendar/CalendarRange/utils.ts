import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
