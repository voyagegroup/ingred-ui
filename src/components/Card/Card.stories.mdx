import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import Card from "./Card";

<Meta title="Components/Layout/Card" component={Card} />

# Card

It is used when we want to express something gathering of information.

It can contains `<Flex />`props & `<Spacer />`props.

<ArgsTable of={Card} />

## Samples

<Canvas>
  <Story
    name="Example"
    args={{
      p: 3,
      children: "contents",
    }}
  >
    {(args) => (
      <div style={{ padding: "10px" }}>
        <Card {...args} />
      </div>
    )}
  </Story>
</Canvas>
