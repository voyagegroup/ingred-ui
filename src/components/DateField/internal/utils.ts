export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
