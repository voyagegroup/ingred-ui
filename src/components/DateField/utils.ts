import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
