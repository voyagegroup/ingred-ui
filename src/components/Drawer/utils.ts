/**
 * 相対値をpx値に変換するヘルパー関数
 * @param value - 変換する値（数値またはvw/vh/%/px等の単位付き文字列）
 * @param direction - Drawerの方向（変換時の基準値決定に使用）
 * @returns px値（数値）
 */
export const convertToPixels = (value: string | number, direction: "left" | "right" | "bottom"): number => {
  if (typeof value === "number") {
    return value; // 数値はpx扱い
  }
  
  // 文字列の場合は単位チェック
  const numericValue = parseFloat(value);
  
  if (value.includes("vw")) {
    return (numericValue / 100) * window.innerWidth;
  }
  if (value.includes("vh")) {
    return (numericValue / 100) * window.innerHeight;
  }
  if (value.includes("%")) {
    // %の場合は親要素基準だが、ここではviewport基準として扱う
    const isHorizontal = direction === "left" || direction === "right";
    return (numericValue / 100) * (isHorizontal ? window.innerWidth : window.innerHeight);
  }
  if (value.includes("px")) {
    return numericValue;
  }
  
  // 単位なしの文字列は警告を出す
  if (!isNaN(numericValue)) {
    console.warn(`Drawer: 文字列サイズには単位を指定してください。"${value}" → "${value}px" を推奨`);
    return numericValue; // 一応数値として扱う
  }
  
  // 無効な値の場合はエラー警告とフォールバック
  console.error(`Drawer: 無効なサイズ値です: "${value}". デフォルト値を使用します。`);
  return 400; // フォールバック
}; 