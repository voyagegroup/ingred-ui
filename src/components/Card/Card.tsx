import styled from "styled-components";
import { Property } from "csstype";

import { flexbox, FlexProps } from "../Flex/Flex";
import { spacer, SpacerProps } from "../../utils/spacer";

export type CardProps = SpacerProps &
  FlexProps & {
    width?: Property.Width;
    maxWidth?: Property.MaxWidth;
    minWidth?: Property.MinWidth;
  };

const Card = styled.div<CardProps>`
  border-radius: ${({ theme }) => theme.radius}px;
  background-color: ${({ theme }) => theme.palette.background.default};
  width: ${({ width }) => width || "auto"};
  min-width: ${({ minWidth }) => minWidth || "auto"};
  max-width: ${({ maxWidth }) => maxWidth || "auto"};
  border: 1px solid ${({ theme }) => theme.palette.gray.light};
  ${spacer}
  ${flexbox}
`;

export default Card;
