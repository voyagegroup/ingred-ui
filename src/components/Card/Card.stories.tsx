import React from "react";
import { Story } from "@storybook/react/types-6-0";
import { Title, Description, ArgsTable, Stories } from "@storybook/addon-docs";
import Card, { CardProps } from "./Card";

export default {
  title: "Components/Layout/Card",
  component: Card,
  args: {
    m: 1,
    p: 3,
    children: "contents",
  },
  parameters: {
    docs: {
      source: { type: "code" },
      page: () => (
        <>
          <Title />
          <Description
            markdown={[
              "It is used when we want to express something gathering of information.",
              "",
              "It can contains `<Flex />`props & `<Spacer />`props.",
            ].join("\n")}
          />
          <ArgsTable of={Card} />
          <Stories includePrimary title="Stories" />
        </>
      ),
    },
  },
};

export const Example: Story<CardProps> = (args) => <Card {...args} />;
