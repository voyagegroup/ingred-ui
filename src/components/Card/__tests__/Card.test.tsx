import * as React from "react";
import "@testing-library/jest-dom";
import { cleanup } from "@testing-library/react";
import Card from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Card component testing", () => {
  afterEach(cleanup);

  test("Card", () => {
    const { asFragment } = renderWithThemeProvider(<Card />);
    expect(asFragment()).toMatchSnapshot();
  });
});
