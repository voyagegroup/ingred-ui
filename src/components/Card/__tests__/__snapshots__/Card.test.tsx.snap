// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component testing Card 1`] = `
<DocumentFragment>
  <div
    class="sc-gsDKAQ fvfjgw"
  />
</DocumentFragment>
`;
